### PR TITLE
Convert record to account

### PR DESCRIPTION
**Pull request by aaronc**

### CODE REVIEW

File: account.go [hunk 1]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+	MinExponent: apd.MinExponent,
+	Traps:       apd.DefaultTraps,
+}
+
+func RecordToAccount(rec Record, genesisTime time.Time) (Account, error) {
+	amount := rec.TotalAmount
+	distTime := rec.StartTime
+	if distTime.IsZero() {
+		distTime = genesisTime
+	}
```
Comment (clevinson) [inline account.go:49]: Is this actually needed? It seems like in parse.go we are always setting this to mainnet, mainnet+1year, or the explicit startTime provided.

File: account.go [hunk 2]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+	numDist := rec.NumMonthlyDistributions
+	if numDist < 1 {
+		numDist = 1
+	}
```
Comment (clevinson) [inline account.go:53]: Might be cleaner to error here. We shouldn't have any 0 values in the spreadsheet.

File: account.go [hunk 3]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+
+func distAmountAndDust(amount apd.Decimal, numDist int) (distAmount apd.Decimal, dust apd.Decimal, err error) {
+	if numDist <= 1 {
+		return amount, dust, nil
+	}
+
+	var numDistDec apd.Decimal
+	numDistDec = *numDistDec.SetInt64(int64(numDist))
+
+	_, err = Dec128Context.Mul(&amount, &amount, tenE6)
```
Comment (clevinson) [inline account.go:123]:
```suggestion
	// convert amount from regen to uregen, so we can perform integral arithmetic on uregen
	_, err = Dec128Context.Mul(&amount, &amount, tenE6)
```

Looks like this is coverting amount from regen => uregen. Can we name the variables more explicitly to make that clear? amtRegen and amtUregen or similar? Alternatively, atleast we sould provide a comment here.

File: account.go [hunk 4]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+	if err != nil {
+		return distAmount, dust, err
+	}
+
+	// each distribution is an integral amount of regen
```
Comment (clevinson) [inline account.go:128]:
```suggestion
	// each distribution is an integral amount of uregen
```

File: account.go [hunk 5]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+	return Account{
+		Address:       rec.Address,
+		Distributions: distributions,
+		TotalRegen:    amount,
+	}, nil
+}
+
+func distAmountAndDust(amount apd.Decimal, numDist int) (distAmount apd.Decimal, dust apd.Decimal, err error) {
+	if numDist <= 1 {
+		return amount, dust, nil
```
Comment (clevinson) [inline account.go:117]: This is new to me, but in golang, does having a variable declared as a return value declare it for usage within the function body? So we don't need declare it in function body?

File: account.go [hunk 6]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+	if err != nil {
+		return distAmount, dust, err
+	}
+
+	_, err = Dec128Context.Rem(&dust, &amount, &numDistDec)
+	if err != nil {
+		return distAmount, dust, err
+	}
+
+	_, err = Dec128Context.Quo(&distAmount, &distAmount, tenE6)
```
Comment (clevinson) [inline account.go:139]:
```suggestion
	// convert distAmount from uregen back to regen
	_, err = Dec128Context.Quo(&distAmount, &distAmount, tenE6)
```

File: account.go [hunk 7]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+	if err != nil {
+		return distAmount, dust, err
+	}
+
+	_, err = Dec128Context.Quo(&dust, &dust, tenE6)
```
Comment (clevinson) [inline account.go:144]:
```suggestion
	// convert dust from uregen back to regen
	_, err = Dec128Context.Quo(&dust, &dust, tenE6)
```

File: account.go [hunk 8]
```diff
@@ -34,6 +34,121 @@ func (d Distribution) String() string {
...
+			Distributions: []Distribution{
+				{
+					Time:  genesisTime,
+					Regen: amount,
+				},
+			},
+		}, nil
+	}
+
+	distAmount, dust, err := distAmountAndDust(amount, numDist)
```
Comment (clevinson) [inline account.go:68]:
```suggestion
	// calculate dust, which represents an uregen-integral remainder, represented as an apd.Decimal of regen
	// from dividing `amount` by `numDist`
	distAmount, dust, err := distAmountAndDust(amount, numDist)
```

File: account_test.go [hunk 9]
```diff
@@ -273,3 +273,315 @@ func TestToCosmosAccount(t *testing.T) {
...
+	twoMonthsBeforeGenesis := genesisTime.Add(-OneMonth * 2)
+
+	tests := []struct {
+		name    string
+		record  Record
+		want    Account
+		wantErr bool
+	}{
+		{
+			"no dists, no start time",
```
Comment (clevinson) [inline account_test.go:308]: We should error on this. All records in the csv will have startTime and atleast 1 distribution.

File: account_test.go [hunk 10]
```diff
@@ -273,3 +273,315 @@ func TestToCosmosAccount(t *testing.T) {
...
+					{
+						Time:  genesisTime,
+						Regen: *five,
+					},
+				},
+			},
+			false,
+		},
+		{
+			"no dists, genesis start time",
```
Comment (clevinson) [inline account_test.go:326]: Same, this should error.

File: account_test.go [hunk 11]
```diff
@@ -273,3 +273,315 @@ func TestToCosmosAccount(t *testing.T) {
...
+		args           args
+		wantDistAmount apd.Decimal
+		wantDust       apd.Decimal
+		wantErr        bool
+	}{
+		{
+			name: "0 dist",
+			args: args{
+				amount:  *ten02,
+				numDist: 0,
```
Comment (clevinson) [inline account_test.go:527]: Should error nor not test this case.

File: account_test.go [hunk 12]
```diff
@@ -273,3 +273,315 @@ func TestToCosmosAccount(t *testing.T) {
...
+		})
+	}
+}
+
+func Test_distAmountAndDust(t *testing.T) {
+	five01, _, err := apd.NewFromString("5.01")
+	require.NoError(t, err)
+	ten02, _, err := apd.NewFromString("10.02")
+	require.NoError(t, err)
+	ten0000002, _, err := apd.NewFromString("10.020001")
```
Comment (clevinson) [inline account_test.go:503]:
```suggestion
	ten020001, _, err := apd.NewFromString("10.020001")
```

File: account_test.go [hunk 13]
```diff
@@ -273,3 +273,315 @@ func TestToCosmosAccount(t *testing.T) {
...
+				numDist: 2,
+			},
+			wantDistAmount: *five01,
+			wantDust:       apd.Decimal{},
+			wantErr:        false,
+		},
+		{
+			name: "2 dist, dust",
+			args: args{
+				amount:  *ten0000002,
```
Comment (clevinson) [inline account_test.go:556]:
```suggestion
				amount:  *ten020001,
```

**Review by clevinson — APPROVED** [review]
Looks good!

Few cosmetic / comment suggestions, and one bigger ask- We're currently not planning on having any records with 0 distributions, or missing `startTime` values in the CSV. So imo it would be preferable to actually error out on all those cases as opposed to handling them as "mainnet" / 1 distribution.

If you have a strong preference for the current logic, then i'm fine to merge. So approving as is as well.

File: account.go (hunk 5 above)
Comment (aaronc) [reply account.go:117]: yeah

**Pull request merged**